### PR TITLE
Builtin functions for linear regression algorithms

### PR DESCRIPTION
**Pull request by kev-inn**
This adds 4 new DML-bodied builtin functions:

  * lm
  * lmDS
  * lmCG
  * lmpredict

These functions are to be used for linear regression.
`lm` calls either `lmDS` or `lmCG` depending on the size of the input matrices. `lmpredict` can be used
to predict the class of a feature vector with the weights usually calculated before by use of `lm`.

This also adds a file used for documentation of builtin functions, which currently only documents those new added builtins.

Beware that lm currently is non-functional and the test cases will fail, because of a namespace conflict when using multiple different DML-bodied builtin functions.

### CODE REVIEW

File: docs/builtins-reference.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
+<!--
+{% comment %}
+Modifications Copyright 2018 Graz University of Technology
```
Comment (mboehm7) [inline docs/builtins-reference.md]: Please use ... 2019.

File: docs/builtins-reference.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  * [Introduction](#introduction)
+  * [DML-bodied builtin functions](#dml-bodied-builtin-functions)
+    * [`lm`-function](#lm-function)
+    * [`lmDS`-function](#lmds-function)
+    * [`lmCG`-function](#lmcg-function)
+    * [`lmpredict`-function](#lmpredict-function)
+    
+# Introduction
+
+Declarative Machine Learning has builtin functions which enable access to both low- and high-level functions
```
Comment (mboehm7) [inline docs/builtins-reference.md]: Maybe "The DML (...) language has ..."

File: docs/builtins-reference.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+The *icpt-argument* can be set to 3 modes:
+ 
+  * 0 = no intercept, no shifting, no rescaling
+  * 1 = add intercept, but neither shift nor rescale X
+  * 2 = add intercept, shift & rescale X columns to mean = 0, variance = 1
+
+### Example
+```r
+X = rand (rows = 50, cols = 10)
+y = rand (rows = 50, cols = 1)
```
Comment (mboehm7) [inline docs/builtins-reference.md]: Please make it more meaningful by using `y = X %*% rand(rows=ncol(X), 1)` (for all examples)

File: scripts/builtin/lm.dml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#-------------------------------------------------------------
+
+m_lm = function(Matrix[Double] X, Matrix[Double] y, Integer icpt = 0, Double reg = 1e-7, Double tol = 1e-7, Integer maxi = 0, Boolean verbose = TRUE)
+    return (Matrix[Double] B) {
+  n = nrow (X);
```
Comment (mboehm7) [inline scripts/builtin/lm.dml]: Please inline the nrow(X) as it is used just once

File: scripts/builtin/lm.dml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  if (n < 2000) {
+    B = lmDS(X, y, icpt, reg, verbose)
+  }
+  else {
+    B = lmCG(X, y, icpt, reg, tol, maxi, verbose)
+  }
```
Comment (mboehm7) [inline scripts/builtin/lm.dml]: Remove the unnecessary curly braces for these one-line conditions

File: scripts/builtin/lmCG.dml [hunk 6]
```diff
@@ -0,0 +1,180 @@
+#-------------------------------------------------------------
+#
+# Copyright 2019 Graz University of Technology
```
Comment (mboehm7) [inline scripts/builtin/lmCG.dml]: Use the modifications-license as the original linregCG was copied from Apache SystemML.

File: scripts/builtin/lmDS.dml [hunk 7]
```diff
@@ -0,0 +1,139 @@
+#-------------------------------------------------------------
+#
+# Copyright 2019 Graz University of Technology
```
Comment (mboehm7) [inline scripts/builtin/lmDS.dml]: Use the modifications-license as the original linregDS was copied from Apache SystemML.

File: src/test/scripts/functions/builtin/lm.R [hunk 8]
```diff
@@ -0,0 +1,26 @@
+#-------------------------------------------------------------
+#
+# Copyright 2018 Graz University of Technology
```
Comment (mboehm7) [inline src/test/scripts/functions/builtin/lm.R:3]: 2019

File: scripts/builtin/lmpredict.dml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#-------------------------------------------------------------
+
+m_lmpredict = function(Matrix[Double] X, Matrix[Double] w) return (Matrix[Double] y) {
+  y = X %*% w
```
Comment (mboehm7) [inline scripts/builtin/lmpredict.dml]: Please extend this to allow for specify intercept as well - in which case you would need to append a column of ones and/or scale / shift.

Comment (mboehm7) [reply scripts/builtin/lmpredict.dml]: For now it's sufficient to support intercept 1 and include a TODO for intercept 2 as this is a longer discussion

**Review by mboehm7 — COMMENTED** [review]
Thanks @kev-inn - this is a good start. Please incorporate the comments and then I'll merge it in.

**Comment by mboehm7** [comment]
@kev-inn Thanks again for catching the issue regarding namespace handling of multiple dml-bodied builtin functions. This issue has been fixed w/ commit 506cc9bae9a6a6135233d85cd84825a3963bb936.

File: scripts/builtin/lmCG.dml (hunk 6 above)
Comment (kev-inn) [reply scripts/builtin/lmCG.dml]: I am sorry could you explain which comment I would have to add instead. Is it the same that as the one inside LinearRegCG.dml?

**Comment by kev-inn** [comment]
> @kev-inn Thanks again for catching the issue regarding namespace handling of multiple dml-bodied builtin functions. This issue has been fixed w/ commit [506cc9b](https://github.com/tugraz-isds/systemds/commit/506cc9bae9a6a6135233d85cd84825a3963bb936).

@mboehm7 I think I detected another case where we get problems now. Calling a builtin in a dml-bodied bultin leads to the error `namespace ./scripts/builtin is not defined`. You can test this with lm once I resolved all the issues.

File: scripts/builtin/lmCG.dml (hunk 6 above)
Comment (kev-inn) [reply scripts/builtin/lmCG.dml]: Found one in DenseBlockDRB, will add it

**Comment by mboehm7** [comment]
ok thanks for catching - then let's finalize this PR and once it's in master, I'll have a look into this issue to avoid redundancy.

File: scripts/builtin/lmCG.dml (hunk 6 above)
Comment (mboehm7) [reply scripts/builtin/lmCG.dml]: exactly

**Comment by mboehm7** [comment]
LGTM - thanks @kev-inn for incorporating the comments. I think that is an excellent start toward higher-level abstractions in SystemDS. During the merge I made only minor formatting changes and I moved the stats computation of lmCG and lmDS in a verbose branch - dead code elimination would have avoided unnecessary operations as well but this makes it more explicit (we can better understand the performance implications on script level).

**Pull request closed without merging**